cmake/templates/filesystem.h.in

```diff
@@ -0,0 +1,44 @@
+/*  @author An Tao

```

Comment:
Ugm... why is this file generated? I may be wrong. But it could be a regular .h.

Reply:
Yes, I agree with you, just add filesystem.h to the lib/inc/drogon/utils path.

Reply:
Well, this is needed for systems where c++17 don't have std::filesystem::path
In cmake, there is a way to check if <filesystem> is available, but none to check if <filesystem> has std::filesystem::path

So I added a try_compile to check that, and the filesystem.h is then generated with or without HAS_STD_FILESYSTEM_PATH defined.
Base on this, either std or boost is used.

Reply:
can we simply use something like `#if __cplusplus >= 201703L || (defined _MSC_VER && _MSC_VER > 1900)` to decide to use std::filesystem or boost::filesystem?
this solution makes application build environment independent with the framework/library build environment.

Reply:
Well, I had 3 solutions:
1. requiring c++20 for std::filesystem::path for all systems
2. requiring c++17 for all systems, except Windows and OS X where c++20 is required.
3. do a compilation test when running cmake

I initially choosed the first solution because it was the simpler one :)
As marty1885 preferred to avoid using Boost on Linux/BSD, and for the following reasons, I then choosed the 3rd solution:
- I wasn't able to find out for which version of __cplusplus it was made available for Apple's Clang
- There is sadly no [feature test](https://en.cppreference.com/w/cpp/feature_test) for std::filesystem::path
- There are perhaps other systems than Windows and OS X that have a partial c++17 implementation with missing filesystem::path.

For Windows I could check `#if defined(_WIN32) && defined(_MSC_VER) && _MSC_VER >= 1920`
But for Mac? Do you have an idea perhaps?

Reply:
I just had the great idea of doing a `cout << __cplusplus` on my old timer Mojave and I now have the magic value: 201707.

It works here on my Mojave with `#if (defined(__APPLE__) && __cplusplus > 201707L) || (defined(_WIN32) && defined(_MSC_VER) && _MSC_VER >= 1920) || (!defined(__APPLE__) && !defined(_WIN32) && __cplusplus >= 201703L)`
I'll check tomorrow at work on a Big Sur and with MSVC 2019.
Sorry, I have no more MSVC 2017 to check.

Of course, this won't guarantee to work 100% of the time if other systems than OS X and MSVC have such partial implementations. Personally, I liked the solution of testing in cmake and generating the header file, what avoid tries & fails.

Reply:
The CXX_STANDARD property of cmake effects the __cplusplus macro. for std::any, std::string_view and std::coroutine , we check them and set the CXX_STANDARD by:
```cmake
check_include_file_cxx(any HAS_ANY)
check_include_file_cxx(string_view HAS_STRING_VIEW)
check_include_file_cxx(coroutine HAS_COROUTINE)
if (HAS_ANY AND HAS_STRING_VIEW AND HAS_COROUTINE)
    set(DROGON_CXX_STANDARD 20)
elseif (HAS_ANY AND HAS_STRING_VIEW)
    set(DROGON_CXX_STANDARD 17)
else ()
    set(DROGON_CXX_STANDARD 14)
endif ()
```
this changes the behavior of any.h, string_view.h and coroutins.h in the /lib/inc/drogon/utils path. and these lines should also be put in the CMakeLists.txt in the project generated by drogon_ctl. 
we should make the headers intalled into system independent from the library build environment, this is good practice, it's good for binary package distributations.
Consider a condition that happened in a user's project with drogon, he used new compiler build and install drogon with c++17 standard, but some other parts of the project required older standard c++14, so he set the  CXX_STANDARD to 14 manually. if we add a macro generated by cmake when building the library, it doesn't work in this condition.
So there is another tip, we should isolate the object of any, string_view in public headers from the inner implementations because they may be different (boost::any in headers and std::any when building the library). this should also be paid attention to the filesystem stuff. If you didn't add new APIs which using filesystem objects, the filesystem.h file is not needed to add to the public headers, if you did, please note the above suggestions, thanks.

Reply:
Damn, you are right. On Big Sur I get a smaller value (201703L) but filesystem::path is available.
Btw, it's still c++17, but complete, with a partial c++20 (missing coroutine, although it's present in experimental/coroutine).
So one cannot rely on __cplusplus on Mac. Furthermore, one can build drogon for a previous version of OSX, so it's the target os version that must be checked instead (__MAC_OS_X_VERSION_MAX_ALLOWED).

About filesystem.h, I made it public since I'm already working in // to the integration in the implementation classes, with the objective to make them after that public.
But if you prefer, I can move it to private headers for now.